lib/src/combined_wrappers/combined_map.dart

```diff
@@ -39,14 +41,56 @@ class CombinedMapView<K, V> extends UnmodifiableMapBase<K, V> {
 
   /// The keys of [this].
   ///
-  /// The returned iterable has efficient `length` and `contains` operations,
-  /// based on [length] and [containsKey] of the individual maps.
+  /// The returned iterable has efficient `contains` operations, based on
+  /// [containsKey] of the individual maps.
+  ///
+  /// The `length` must do deduplication and thus is not optimized.
   ///
   /// The order of iteration is defined by the individual `Map` implementations,
   /// but must be consistent between changes to the maps.
   ///
   /// Unlike most [Map] implementations, modifying an individual map while
   /// iterating the keys will _sometimes_ throw. This behavior may change in
   /// the future.
-  Iterable<K> get keys => CombinedIterableView<K>(_maps.map((m) => m.keys));
+  Iterable<K> get keys => _LazyDeduplicatingIterableView(
+      CombinedIterableView(_maps.map((m) => m.keys)));
```

Comment:
Is the `CombinedIterableView` buying us much over `expand`? Does it get us more than an `isEmpty` optimization?

I don't think we need `_LazyDeduplicatingIterableView` if we use `sync*` since that will already re-invoke the function on each retrieval of the iterator. Not sure how much we want to try to micro-optimize.

```dart
Iterable<K> get keys sync* {
  final seen = <K>{};
  yield* _maps.expand((m) => m.keys).where(seen.add);
}
```

Reply:
Hmm, the `contains` optimization is a little more compelling...

Reply:
I think the contains optimization is worth it for something in a "core" lib like this.